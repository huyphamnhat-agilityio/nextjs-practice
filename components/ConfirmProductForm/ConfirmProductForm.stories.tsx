import type { Meta, StoryObj } from "@storybook/react";
import { useArgs } from "@storybook/preview-api";

// Components
import { ConfirmProductForm } from "..";

const meta: Meta<typeof ConfirmProductForm> = {
  title: "Components/ConfirmProductForm",
  component: ConfirmProductForm,
  argTypes: {
    id: {
      description: "Product ID",
    },
    isOpen: {
      description: "Whether the form is open or not",
    },
    onClose: {
      description: "Callback function to close the form",
    },
    onOpen: {
      description: "Callback function to open the form",
    },
    onOpenChange: {
      description: "Callback function to handle form opening/closing changes",
    },
  },
  parameters: {
    controls: {
      expanded: true,
    },
  },
};

export default meta;

type Story = StoryObj<typeof ConfirmProductForm>;

export const Primary: Story = {
  args: {
    isOpen: true,
    id: "0",
    onOpen: () => {},
    onOpenChange: () => {},
  },
  render: function Render(args) {
    const [, updateArgs] = useArgs<typeof args>();

    const onClose = () => {
      updateArgs({ isOpen: false });
    };

    return <ConfirmProductForm {...args} onClose={onClose} />;
  },
};
