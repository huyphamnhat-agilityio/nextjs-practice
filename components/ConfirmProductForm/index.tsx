"use client";
import {
  Modal,
  ModalBody,
  ModalContent,
  ModalFooter,
  ModalHeader,
} from "@nextui-org/react";
import toast from "react-hot-toast";
import { useFormStatus } from "react-dom";

// Component
import { Button, Input } from "../common";

// Services
import { deleteProduct } from "@/lib";

export type ConfirmProductFormBodyProps = {
  onClose: () => void;
  id: string;
};

const ConfirmProductFormBody = ({
  onClose,
  id,
}: ConfirmProductFormBodyProps) => {
  const { pending } = useFormStatus();

  return (
    <>
      <ModalBody className="flex">
        Do you want to delete this course? This action cannot be undone.
        <Input name="id" value={id} className="hidden" />
      </ModalBody>
      <ModalFooter>
        <Button
          color="danger"
          variant="flat"
          isDisabled={pending}
          onPress={onClose}
        >
          Cancel
        </Button>
        <Button
          color="primary"
          className="text-white"
          type="submit"
          isDisabled={pending}
        >
          Delete
        </Button>
      </ModalFooter>
    </>
  );
};

export type ConfirmProductFormProps = {
  isOpen: boolean;
  onOpen: () => void;
  onOpenChange: () => void;
  onClose: () => void;
  id: string;
};

const ConfirmProductForm = ({
  id,
  isOpen,
  onClose,
  onOpenChange,
}: ConfirmProductFormProps) => {
  const deleteAction = async (payload: FormData) => {
    try {
      const result = await deleteProduct(payload);

      if (result.message) toast.success(result.message);
    } catch (error) {
      if (error instanceof Error) toast.error(error.message);
    }

    onClose();
  };
  return (
    <Modal isOpen={isOpen} onOpenChange={onOpenChange} scrollBehavior="outside">
      <ModalContent>
        <ModalHeader className="text-2xl">Delete course</ModalHeader>
        <form action={deleteAction}>
          <ConfirmProductFormBody id={id} onClose={onClose} />
        </form>
      </ModalContent>
    </Modal>
  );
};

export default ConfirmProductForm;
