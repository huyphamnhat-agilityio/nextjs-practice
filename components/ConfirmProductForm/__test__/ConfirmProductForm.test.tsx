import { render, screen } from "@testing-library/react";
import { useFormStatus } from "react-dom";
import { usePathname, useSearchParams } from "next/navigation";
import userEvent from "@testing-library/user-event";

import ConfirmProductForm, { ConfirmProductFormProps } from "..";

// Mock necessary hooks
jest.mock("react-dom", () => ({
  ...jest.requireActual("react-dom"),
  useFormStatus: jest.fn(),
}));

jest.mock("next/navigation", () => ({
  usePathname: jest.fn(),
  useSearchParams: jest.fn(),
}));

jest.mock("../../../lib/product.ts", () => ({
  deleteProduct: jest.fn(),
}));
describe("ConfirmProductForm test cases", () => {
  const onClose = jest.fn();
  const onOpenChange = jest.fn();
  const onOpen = jest.fn();

  beforeAll(() => {
    window.addEventListener("submit", (e) => {
      e.preventDefault();
    });
  });

  beforeEach(() => {
    (useFormStatus as jest.Mock).mockReturnValue({ pending: false });
    (usePathname as jest.Mock).mockReturnValue("/");
    (useSearchParams as jest.Mock).mockReturnValue(new URLSearchParams());
  });

  const setup = (props: ConfirmProductFormProps) =>
    render(<ConfirmProductForm {...props} />);

  const mockProps: ConfirmProductFormProps = {
    id: "1",
    isOpen: true,
    onClose: onClose,
    onOpen: onOpen,
    onOpenChange: onOpenChange,
  };

  it("should renders correctly", () => {
    const { asFragment } = setup(mockProps);

    expect(asFragment()).toMatchSnapshot();
  });

  it("disables the Delete button when pending", () => {
    (useFormStatus as jest.Mock).mockReturnValue({ pending: true });

    setup(mockProps);

    const submitBtn = screen.getByRole("button", {
      name: /delete/i,
    });

    expect(submitBtn).toBeDisabled();
  });

  it("triggers form submission with the correct action", async () => {
    const { asFragment } = setup(mockProps);

    const submitBtn = screen.getByRole("button", {
      name: /delete/i,
    });

    await userEvent.click(submitBtn);

    expect(asFragment()).toMatchSnapshot();
  });
});
