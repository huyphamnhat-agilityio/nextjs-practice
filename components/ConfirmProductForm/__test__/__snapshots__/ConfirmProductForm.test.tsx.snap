// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmProductForm test cases should renders correctly 1`] = `<DocumentFragment />`;

exports[`ConfirmProductForm test cases triggers form submission with the correct action 1`] = `<DocumentFragment />`;
